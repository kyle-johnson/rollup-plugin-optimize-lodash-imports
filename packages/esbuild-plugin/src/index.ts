import fs from "fs";

import * as acorn from "acorn";
import { Plugin } from "esbuild";
import {
  ParseFunction,
  transform,
  UNCHANGED,
} from "@optimize-lodash/transform";

const wrappedParse: ParseFunction = (code) =>
  acorn.parse(code, { ecmaVersion: "latest", sourceType: "module" });

export type PluginOptions = {
  useLodashEs?: true;
  appendDotJs?: boolean;
};

// TODO: filter https://golang.org/pkg/regexp/
export function lodashOptimizeImports({
  useLodashEs,
  appendDotJs = true,
}: PluginOptions = {}): Plugin {
  const cache = new Map<
    string,
    { input: string; output: string | UNCHANGED }
  >();

  return {
    name: "lodash-optimize-imports",
    setup(build) {
      build.onLoad({ filter: /.(js|ts|jsx|tsx)$/ }, async ({ path }) => {
        const input = await fs.promises.readFile(path, "utf8");
        const cached = cache.get(path); // TODO: unit test the cache

        if (cached && input === cached.input && cached.output === UNCHANGED) {
          return;
        }

        const result = transform({
          code: input,
          id: path,
          parse: wrappedParse,
          useLodashEs,
          appendDotJs,
        });
        if (result === UNCHANGED) {
          cache.set(path, { input, output: UNCHANGED });
          return;
        }

        const output = result.code;
        cache.set(path, { input, output });
        return { contents: output };
      });
    },
  };
}
