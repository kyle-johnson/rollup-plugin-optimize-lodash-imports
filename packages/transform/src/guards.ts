import type {
  BaseNode,
  ImportDeclaration,
  ImportSpecifier,
  Program,
} from "estree";

export function isImportDeclaration(node: BaseNode): node is ImportDeclaration {
  return node.type === "ImportDeclaration";
}

export function isProgram(node: BaseNode): node is Program {
  return node.type === "Program";
}

export function isImportSpecifierArray(
  items: ImportDeclaration["specifiers"],
): items is Array<ImportSpecifier> {
  return items.every((item) => item.type === "ImportSpecifier");
}
