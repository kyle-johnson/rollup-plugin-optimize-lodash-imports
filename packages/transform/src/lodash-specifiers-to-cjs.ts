import type { ImportSpecifier } from "estree";

/**
 * Turns a generic lodash import into a specific import using the CommonJS
 * lodash package.
 *
 * @param base "lodash" or "lodash/fp"
 * @param specifiers from an AST; assumes they are all ImportSpecifiers
 * @param appendDotJs optional, default is true; adds '.js' to the end of imports
 */
export function lodashSpecifiersToCjs(
  base: string,
  specifiers: Array<ImportSpecifier>,
  appendDotJs = true
): Array<string> {
  return specifiers.map(
    ({ imported, local }) =>
      `import ${
        imported.name !== local.name ? local.name : imported.name
      } from "${base}/${imported.name}${appendDotJs ? ".js" : ""}";`
  );
}
