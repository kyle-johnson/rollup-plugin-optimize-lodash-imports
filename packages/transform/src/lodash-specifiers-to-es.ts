import type { ImportSpecifier } from "estree";

/**
 * Turns a generic lodash import into a specific import referencing the "lodash-es"
 * pacakge. Note that lodash-es cannot be imported from CommonJS.
 *
 * @param base "lodash" or "lodash/fp"
 * @param specifiers from an AST; assumes they are all ImportSpecifiers
 */
export function lodashSpecifiersToEs(
  base: string,
  specifiers: Array<ImportSpecifier>,
): Array<string> {
  const isFp = base.endsWith("fp");
  return specifiers.map(
    ({ imported, local }) =>
      `import { ${
        imported.name !== local.name
          ? imported.name + " as " + local.name
          : local.name
      } } from "lodash-es${isFp ? "/fp" : ""}";`,
  );
}
