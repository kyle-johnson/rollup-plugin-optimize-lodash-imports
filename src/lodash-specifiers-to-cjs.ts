import type { ImportSpecifier } from "estree";

/**
 * Turns a generic lodash import into a specific import using the CJS
 * lodash package.
 *
 * @param base "lodash" or "lodash/fp"
 * @param specifiers from an AST, assumes they are all ImportSpecifiers
 */
export function lodashSpecifiersToCjs(
  base: string,
  specifiers: Array<ImportSpecifier>
): Array<string> {
  return specifiers.map(
    ({ imported, local }) =>
      `import ${
        imported.name !== local.name ? local.name : imported.name
      } from "${base}/${imported.name}";`
  );
}
