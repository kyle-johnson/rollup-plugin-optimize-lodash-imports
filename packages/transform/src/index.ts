import type { Node } from "acorn";
import MagicString, { SourceMap } from "magic-string";
import { walk } from "estree-walker";

import {
  isImportDeclaration,
  isImportSpecifierArray,
  isProgram,
} from "./guards";
import { lodashSpecifiersToEs } from "./lodash-specifiers-to-es";
import { lodashSpecifiersToCjs } from "./lodash-specifiers-to-cjs";

// acorn adds these
declare module "estree" {
  interface BaseNodeWithoutComments {
    // added by acorn
    start: number;
    end: number;
  }
}

export type UNCHANGED = null;
export const UNCHANGED = null;

export interface CodeWithSourcemap {
  code: string;
  map: SourceMap;
}
export type ParseFunction = (code: string) => Node;
export type WarnFunction = (message: string) => void;

export function transform({
  code,
  id,
  parse,
  warn,
  useLodashEs,
  appendDotJs = true,
}: {
  code: string;
  id: string;
  parse: ParseFunction;
  warn?: WarnFunction;
  useLodashEs?: true;
  appendDotJs?: boolean;
}): CodeWithSourcemap | UNCHANGED {
  // before parsing, check if we can skip the whole file
  if (!code.includes("lodash")) {
    return UNCHANGED;
  }

  let ast;
  try {
    ast = parse(code);
  } catch (error) {
    // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
    (error as Error).message += ` in ${id}`;
    throw error;
  }

  // source map generation
  let magicString: MagicString | undefined;

  walk(ast, {
    enter(node) {
      // top-level node; we need to walk its children to find imports
      if (isProgram(node)) {
        return;
      }

      // skip any nodes that aren't imports (this skips most everything)
      if (!isImportDeclaration(node)) {
        this.skip();
        return;
      }

      // narrow-in on lodash imports we care about
      if (node.source.value !== "lodash" && node.source.value !== "lodash/fp") {
        this.skip();
        return;
      }

      // transform specific "lodash" and "lodash/fp" imports such as:
      // import { isNil } from "lodash";
      if (isImportSpecifierArray(node.specifiers)) {
        magicString = magicString ?? new MagicString(code);

        // modify
        const imports = useLodashEs
          ? lodashSpecifiersToEs(node.source.value, node.specifiers)
          : lodashSpecifiersToCjs(
              node.source.value,
              node.specifiers,
              appendDotJs
            );

        // write
        magicString.overwrite(node.start, node.end, imports.join("\n"));

        // no need to dig deeper
        this.skip();
      } else if (warn !== undefined) {
        // help end-users benefit from this plugin (this behavior differs from
        // babel-plugin-lodash which does optimize non-specific imports)
        warn(
          `Detected a default lodash or lodash/fp import within ${id} on line ${
            node.loc?.start?.line ?? "unknown"
          }.\nThis import cannot be optimized by optimize-lodash-imports.`
        );
      }
    },
  });

  if (!magicString) {
    return UNCHANGED;
  }

  return {
    code: magicString.toString(),
    map: magicString.generateMap({
      file: id,
      includeContent: true,
      hires: true,
    }),
  };
}
