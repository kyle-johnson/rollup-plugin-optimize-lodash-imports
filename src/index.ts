import type { Plugin } from "rollup";
import { createFilter, FilterPattern } from "@rollup/pluginutils";
import MagicString from "magic-string";
import { walk } from "estree-walker";

import {
  isImportDeclaration,
  isImportSpecifierArray,
  isProgram,
} from "./guards";
import { lodashSpecifiersToCjs } from "./lodash-specifiers-to-cjs";
import { lodashSpecifiersToEs } from "./lodash-specifiers-to-es";

declare module "estree" {
  interface BaseNodeWithoutComments {
    // added by acorn
    start: number;
    end: number;
  }
}

export type OptimizeLodashOptions = {
  /**
   * A minimatch pattern, or array of patterns, of files that should be
   * processed by this plugin (if omitted, all files are included by default)
   */
  include?: FilterPattern;
  /**
   * Files that should be excluded, if `include` is otherwise too permissive.
   */
  exclude?: FilterPattern;

  /**
   * Changes *all* lodash imports (but not lodash/fp imports!) to 'lodash-es' imports.
   * Don't use this for CommonJS outputs, the plugin will error should you do so.
   */
  useLodashEs?: boolean;
};

const UNCHANGED = null;

/**
 * Converts lodash imports to be specific, enabling better tree-shaking:
 *
 * `import { isNil } from "lodash";` -> `import { isNil } from "lodash/isNil";`
 *
 * Note that only specific named imports are supported, unlike babel-plugin-lodash. For example,
 * this plugin will print a warning for this import and make no changes to the import:
 *
 * `import _ from "lodash";`
 *
 * Optionally, set `useLodashEs` to true and `lodash` imports will be converted to `lodash-es`
 * imports. Note that it's up to user to include the `lodash-es` module and ensure the output
 * is set to some form of `es` (other output formats will error). An example:
 *
 * `import { isNil } from "lodash";` -> `import { isNil } from "lodash-es";`
 *
 * @param include files/globs to include with this plugin (optional)
 * @param exclude files/globs to exclude from this plugin (optional)
 * @param useLodashEs set `true` to convert imports to use "lodash-es" (optional; default false)
 */
export default function optimizeLodashImports({
  include,
  exclude,
  useLodashEs,
}: OptimizeLodashOptions = {}): Plugin & Required<Pick<Plugin, "transform">> {
  const filter = createFilter(include, exclude);

  return {
    name: "optimize-lodash-imports",
    outputOptions(options) {
      if (useLodashEs && options.format !== "es") {
        this.error(
          `'useLodashEs' is true but the output format is not 'es', it's ${
            options.format ?? "undefined"
          }`
        );
      }
      return UNCHANGED;
    },
    transform(code, id) {
      const warn = this.warn.bind(this);

      // sometimes we can skip the whole file
      if (!filter(id) || !code.includes("lodash")) {
        return UNCHANGED;
      }

      let ast;
      try {
        ast = this.parse(code);
      } catch (error) {
        // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
        error.message += ` in ${id}`;
        throw error;
      }

      // easy source map generation
      let magicString: MagicString | undefined;

      walk(ast, {
        enter(node) {
          // top-level node; we need to walk its children to find imports
          if (isProgram(node)) {
            return;
          }

          // skip any nodes that aren't imports (this skips most everything)
          if (!isImportDeclaration(node)) {
            this.skip();
            return;
          }

          // narrow-in on lodash imports we care about
          if (
            node.source.value !== "lodash" &&
            node.source.value !== "lodash/fp"
          ) {
            this.skip();
            return;
          }

          // transform specific "lodash" and "lodash/fp" imports such as:
          // import { isNil } from "lodash";
          if (isImportSpecifierArray(node.specifiers)) {
            magicString = magicString ?? new MagicString(code);

            // modify
            const imports = useLodashEs
              ? lodashSpecifiersToEs(node.source.value, node.specifiers)
              : lodashSpecifiersToCjs(node.source.value, node.specifiers);

            // write
            magicString.overwrite(node.start, node.end, imports.join("\n"));

            // no need to dig deeper
            this.skip();
          } else {
            // help end-users benefit from this plugin (this behavior differs from
            // babel-plugin-lodash which does optimize non-specific imports)
            warn(
              `Detected a default lodash or lodash/fp import within ${id} on line ${
                node.loc?.start?.line ?? "unknown"
              }.\nThis import cannot be optimized by optimize-lodash-imports.`
            );
          }
        },
      });

      if (!magicString) {
        return UNCHANGED;
      }

      return {
        code: magicString.toString(),
        map: magicString.generateMap({
          file: id,
          includeContent: true,
          hires: true,
        }),
      };
    },
  };
}
