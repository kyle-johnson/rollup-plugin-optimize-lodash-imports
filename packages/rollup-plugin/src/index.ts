import type { Plugin } from "rollup";
import { createFilter, FilterPattern } from "@rollup/pluginutils";
import { transform as lodashTransform } from "@optimize-lodash/transform";

export type OptimizeLodashOptions = {
  /**
   * A minimatch pattern, or array of patterns, of files that should be
   * processed by this plugin (if omitted, all files are included by default)
   */
  include?: FilterPattern;
  /**
   * Files that should be excluded, if `include` is otherwise too permissive.
   */
  exclude?: FilterPattern;

  /**
   * Changes *all* lodash imports (but not lodash/fp imports!) to 'lodash-es' imports.
   * Don't use this for CommonJS outputs, the plugin will error should you do so.
   */
  useLodashEs?: true;
};

const UNCHANGED = null;

/**
 * Converts lodash imports to be specific, enabling better tree-shaking:
 *
 * `import { isNil } from "lodash";` -> `import { isNil } from "lodash/isNil";`
 *
 * Note that only specific named imports are supported, unlike babel-plugin-lodash. For example,
 * this plugin will print a warning for this import and make no changes to the import:
 *
 * `import _ from "lodash";`
 *
 * Optionally, set `useLodashEs` to true and `lodash` imports will be converted to `lodash-es`
 * imports. Note that it's up to user to include the `lodash-es` module and ensure the output
 * is set to some form of `es` (other output formats will error). An example:
 *
 * `import { isNil } from "lodash";` -> `import { isNil } from "lodash-es";`
 *
 * @param include files/globs to include with this plugin (optional)
 * @param exclude files/globs to exclude from this plugin (optional)
 * @param useLodashEs set `true` to convert imports to use "lodash-es" (optional; default false)
 */
export function optimizeLodashImports({
  include,
  exclude,
  useLodashEs,
}: OptimizeLodashOptions = {}): Plugin & Required<Pick<Plugin, "transform">> {
  const filter = createFilter(include, exclude);

  return {
    name: "optimize-lodash-imports",
    outputOptions(options) {
      if (useLodashEs && options.format !== "es") {
        this.error(
          `'useLodashEs' is true but the output format is not 'es', it's ${
            options.format ?? "undefined"
          }`
        );
      }
      return UNCHANGED;
    },
    transform(code, id) {
      const warn = this.warn.bind(this);
      const parse = this.parse.bind(this);

      // honor include/exclude
      if (!filter(id)) {
        return UNCHANGED;
      }

      return lodashTransform({ code, id, parse, warn, useLodashEs });
    },
  };
}
