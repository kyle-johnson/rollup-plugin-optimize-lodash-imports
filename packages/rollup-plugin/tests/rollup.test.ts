/**
 * Basic end-to-end integration tests with Rollup.
 *
 * No mocks/stubs, and output is validated.
 */
import { OutputOptions, rollup } from "rollup";

import { optimizeLodashImports, OptimizeLodashOptions } from "../src";

// quiet auto-external warnings with a simple auto-external test
const external = (id: string): boolean => /^[^./]/.test(id);

/**
 * Simplify tests by running rollup and generating output. Errors within rollup will
 * throw. The lodash plugin can be disabled by passing `false` in place of options.
 *
 * @param input input filename
 * @param pluginOptions set `false` to disable the plugin
 * @param rollupOutputFormat cjs, es, etc
 */
const wrapperRollupGenerate = async (
  input: string,
  pluginOptions: OptimizeLodashOptions | false,
  rollupOutputFormat: OutputOptions["format"] = "cjs"
): Promise<string> => {
  const bundle = await rollup({
    input,
    external,
    plugins:
      pluginOptions !== false ? [optimizeLodashImports(pluginOptions)] : [],
  });
  const { output } = await bundle.generate({
    format: rollupOutputFormat,
    validate: true,
  });
  return output[0].code;
};

const STANDARD_AND_FP = `${__dirname}/fixtures/standard-and-fp.js`;

describe("rollup", () => {
  test("setting `useLodashEs` to true with output format `cjs` throws", async () => {
    await expect(
      wrapperRollupGenerate(STANDARD_AND_FP, { useLodashEs: true }, "cjs")
    ).rejects.toThrowErrorMatchingInlineSnapshot(
      `"'useLodashEs' is true but the output format is not 'es', it's cjs"`
    );
  });

  describe("named lodash and lodash/fp imports", () => {
    test("without plugin", async () => {
      expect.assertions(3);
      const code = await wrapperRollupGenerate(STANDARD_AND_FP, false, "cjs");

      // ensure all imports remained untouched
      expect(code).toMatch(/["']lodash["']/g);
      expect(code).toMatch(/["']lodash\/fp["']/g);

      // full snapshot
      expect(code).toMatchSnapshot();
    });

    test.each<[OutputOptions["format"]]>([["cjs"], ["es"]])(
      "with plugin & %s output",
      async (outputFormat) => {
        expect.assertions(3);
        const code = await wrapperRollupGenerate(
          STANDARD_AND_FP,
          {},
          outputFormat
        );

        // ensure all imports became more specific
        expect(code).not.toMatch(/["']lodash["']/g);
        expect(code).not.toMatch(/["']lodash\/fp["']/g);

        // full snapshot
        expect(code).toMatchSnapshot();
      }
    );

    test("with plugin, ES output, & useLodashEs", async () => {
      expect.assertions(4);
      const code = await wrapperRollupGenerate(
        STANDARD_AND_FP,
        { useLodashEs: true },
        "es"
      );

      // ensure all imports became more specific
      expect(code).not.toMatch(/["']lodash["']/g);
      expect(code).not.toMatch(/["']lodash\/fp["']/g);
      expect(code).toMatch(/["']lodash-es["']/g);

      // full snapshot
      expect(code).toMatchSnapshot();
    });
  });
});
