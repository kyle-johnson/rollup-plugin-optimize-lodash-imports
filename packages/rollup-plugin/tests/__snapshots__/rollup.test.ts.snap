// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rollup named lodash and lodash/fp imports with plugin & cjs output 1`] = `
"'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var isNil = require('lodash/isNil.js');
var negate = require('lodash/negate.js');
var every = require('lodash/fp/every.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var isNil__default = /*#__PURE__*/_interopDefaultLegacy(isNil);
var negate__default = /*#__PURE__*/_interopDefaultLegacy(negate);
var every__default = /*#__PURE__*/_interopDefaultLegacy(every);

const everyNonNil = every__default[\\"default\\"](negate__default[\\"default\\"](isNil__default[\\"default\\"]));

function isNonNilArray(input) {
  return Array.isArray(input) && everyNonNil(input);
}

exports.isNonNilArray = isNonNilArray;
"
`;

exports[`rollup named lodash and lodash/fp imports with plugin & es output 1`] = `
"import isNil from 'lodash/isNil.js';
import negate from 'lodash/negate.js';
import every from 'lodash/fp/every.js';

const everyNonNil = every(negate(isNil));

function isNonNilArray(input) {
  return Array.isArray(input) && everyNonNil(input);
}

export { isNonNilArray };
"
`;

exports[`rollup named lodash and lodash/fp imports with plugin, ES output, & useLodashEs 1`] = `
"import { negate, isNil } from 'lodash-es';
import { every } from 'lodash-es/fp';

const everyNonNil = every(negate(isNil));

function isNonNilArray(input) {
  return Array.isArray(input) && everyNonNil(input);
}

export { isNonNilArray };
"
`;

exports[`rollup named lodash and lodash/fp imports without plugin 1`] = `
"'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var lodash = require('lodash');
var fp = require('lodash/fp');

const everyNonNil = fp.every(lodash.negate(lodash.isNil));

function isNonNilArray(input) {
  return Array.isArray(input) && everyNonNil(input);
}

exports.isNonNilArray = isNonNilArray;
"
`;
